src/goto-programs/goto_coverage.cpp

```diff
@@ -38,64 +62,666 @@ void goto_coveraget::add_false_asserts(goto_functionst &goto_functions)
       {
         if (it->is_end_function())
         {
-          insert_false_assert(goto_program, it);
+          // insert an assert(0) as instrumentation BEFORE each instruction
+          insert_assert(goto_program, it, gen_false_expr());
           continue;
         }
 
         if ((!is_true(it->guard) && it->is_goto()) || it->is_target())
         {
           it++; // add an assertion behind the instruciton
-          insert_false_assert(goto_program, it);
+          insert_assert(goto_program, it, gen_false_expr());
           continue;
         }
       }
 
       goto_programt::targett it = goto_program.instructions.begin();
-      insert_false_assert(goto_program, it);
+      insert_assert(goto_program, it, gen_false_expr());
     }
 }
 
-void goto_coveraget::insert_false_assert(
+void goto_coveraget::insert_assert(
   goto_programt &goto_program,
-  goto_programt::targett &it)
+  goto_programt::targett &it,
+  const expr2tc &guard)
 {
   goto_programt::targett t = goto_program.insert(it);
   t->type = ASSERT;
-  t->guard = gen_false_expr();
+  t->guard = guard;
   t->location = it->location;
-  t->location.property("assertion");
-  t->location.comment("Instrumentation ASSERT(0) Added");
+  t->location.property("instrumented assertion");
+  t->location.comment(from_expr(ns, "", guard));
   t->location.user_provided(true);
   it = ++t;
-  total_instrument++;
 }
 
 int goto_coveraget::get_total_instrument() const
 {
+  int total_instrument = 0;
+  forall_goto_functions (f_it, goto_functions)
+    if (f_it->second.body_available && f_it->first != "__ESBMC_main")
+    {
+      const goto_programt &goto_program = f_it->second.body;
+      forall_goto_program_instructions (it, goto_program)
+      {
+        if (
+          it->is_assert() &&
+          it->location.property().as_string() == "instrumented assertion" &&
+          it->location.user_provided() == true)
+        {
+          total_instrument++;
+        }
+      }
+    }
   return total_instrument;
 }
 
-// Obtain total assertion instances in goto level via goto-unwind api
+// Count the total assertion instances in goto level via goto-unwind api
 // run the algorithm on the copy of the original goto program
-void goto_coveraget::gen_assert_instance(goto_functionst goto_functions)
+int goto_coveraget::get_total_assert_instance() const
 {
   // 1. execute goto uniwnd
+  int total_assert_instance = 0;
   bounded_loop_unroller unwind_loops;
   unwind_loops.run(goto_functions);
   // 2. calculate the number of assertion instance
+  forall_goto_functions (f_it, goto_functions)
```

Comment:
This entire loop is basically copy of goto_coveraget::get_total_instrument(), right? Could you unify them?

Reply:
Will change